include/envoy/network/connection.h

```diff
@@ -205,10 +205,50 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    * @return boolean telling if the connection is currently above the high watermark.
    */
   virtual bool aboveHighWatermark() const PURE;
+
+  enum class PostIoAction { Close, KeepOpen };
+
+  struct IoResult {
+    PostIoAction action_;
+    uint64_t bytes_processed_;
+  };
+};
+
+class TransportSecurityCallbacks {
+public:
+  virtual ~TransportSecurityCallbacks(){};
+
+  virtual const Connection& connection() const PURE;
+
+  virtual void raiseEvent(ConnectionEvent event) PURE;
+
+  virtual bool shouldDrainReadBuffer() PURE;
+
+  virtual void setReadBufferReady() PURE;
+
+  virtual int fd() PURE;
+
+  virtual Buffer::Instance& readBuffer() PURE;
+
+  virtual Buffer::Instance& writeBuffer() PURE;
```

Comment:
These should all definitely be commented up - I assume you were just waiting until the APIs were hashed out?

Reply:
Yes, this is still WIP and of course I'll comment up once we agreed on API.